Modules/_threadmodule.c

```diff
@@ -1148,20 +1091,18 @@ static PyObject *
 rlock_release_save(PyObject *op, PyObject *Py_UNUSED(ignored))
 {
     rlockobject *self = (rlockobject*)op;
-    PyThread_ident_t owner;
-    unsigned long count;
 
-    if (self->rlock_count == 0) {
+    PyThread_ident_t owner = self->lock.thread;
+    size_t count = self->lock.level + 1;
+
+    if (_PyRecursiveMutex_TryUnlock(&self->lock) < 0)
+    {
         PyErr_SetString(PyExc_RuntimeError,
                         "cannot release un-acquired lock");
         return NULL;
     }
-
-    owner = self->rlock_owner;
-    count = self->rlock_count;
-    self->rlock_count = 0;
-    _Py_atomic_store_ullong_relaxed(&self->rlock_owner, 0);
-    PyThread_release_lock(self->rlock_lock);
+    self->lock.thread = 0;
+    self->lock.level = 0;
```

Comment:
I don't think this implementation works correctly. `rlock_release_save` is called by `Condition` when it's about to wait. It needs to release the lock (even if it's acquired multiple times) and save the level so that when it wakes up and can attempt to reacquire the lock with the same level. The problem here is:

* `_PyRecursiveMutex_TryUnlock` may not unlock the lock (depending on the level)
* If it does unlock (because level was zero), then the assignments below are unsafe. Someone else might now own the lock when we try to set `thread` and `level` to 0.
* If it does not unlock (because level > 0), then we never actually unlock the underlying mutex nor wakeup waiting threads.

Here's how I think this should be written (actually tested this time):

```c
    rlockobject *self = (rlockobject*)op;

    if (!_PyRecursiveMutex_IsLockedByCurrentThread(&self->lock)) {
        PyErr_SetString(PyExc_RuntimeError,
                        "cannot release un-acquired lock");
        return NULL;
    }

    PyThread_ident_t owner = self->lock.thread;
    size_t count = self->lock.level + 1;
    self->lock.level = 0;  // ensure the unlock releases the lock
    _PyRecursiveMutex_Unlock(&self->lock);
    return Py_BuildValue("k" Py_PARSE_THREAD_IDENT_T, count, owner);
```

Reply:
Thanks for the explanation! I fixed it as you suggested above